server/src/service.rs

```diff
@@ -99,9 +99,27 @@ impl Service {
     #[zbus(out_args("locked", "prompt"))]
     pub async fn lock(
         &mut self,
-        _objects: Vec<OwnedObjectPath>,
+        objects: Vec<OwnedObjectPath>,
+        #[zbus(signal_emitter)] signal_emitter: SignalEmitter<'_>,
     ) -> Result<(Vec<OwnedObjectPath>, ObjectPath), ServiceError> {
-        todo!()
+        let mut locked: Vec<OwnedObjectPath> = Vec::new();
+        let prompt = ObjectPath::default();
+        let collections = self.collections.lock().await;
+
+        for object in &objects {
+            for collection in collections.iter() {
+                if object == collection.path() || collection.items().await.contains(object) {
```

Comment:
So the first part makes sense, but the second one, not sure about it :) because you would lock a collection just because the user locked a single item. That doesn't seem correct.

Reply:
Yeah :) But, this is what gnome-keyring-daemon does. I also think this is wrong. What should I do here?

Reply:
I said before, gnome-keyring-daemon is not a reference implementation.

There should be a clear difference between case when you lock/unlock the whole collection or when you lock / unlock an item in the collection.

Reply:
Thanks for the reply.

My priority is GNOME :) We don't have this in Seahorse and KeyRack. I'm not sure about KDE side.

Reply:
Sure, but the goal is not to re-write the same crap otherwise we can just keep gnome-keyring-daemon.

Reply:
So, we are planning to deprecate gnome-keyring-daemon because its implementation is really complex (to maintain and to add new features: TPM and homed related support). And we want the new daemon to support the latest security requirements. Other than that gnome-keyring-daemon works ok (can keep up with existing client requirements). So, I don't think that we should dismiss/ignore its inner workings, at the least the Secret Service implementation part :)

I have modified this change to support locking an item without locking the whole collection.